File: Queue.c
#include <stdio.h>
#include <stdlib.h>

// Node structure
typedef struct node
{
    int data;
    struct node *next;
} NODE, *PNODE;

// Queue structure
typedef struct queue
{
    PNODE First;
    int iCount;
} QUEUE;

// Function prototypes
void Initialize(QUEUE *q);
void Enqueue(QUEUE *q, int no);
int Dequeue(QUEUE *q);
void Display(QUEUE *q);
int Count(QUEUE *q);

void Initialize(QUEUE *q)
{
    q->First = NULL;
    q->iCount = 0;
}

void Enqueue(QUEUE *q, int no) // InsertLast
{
    PNODE newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->next = NULL;

    if (q->First == NULL)
    {
        q->First = newn;
    }
    else
    {
        PNODE temp = q->First;
        while (temp->next != NULL)
        {
            temp = temp->next;
        }
        temp->next = newn;
    }
    q->iCount++;
}

int Dequeue(QUEUE *q) // DeleteFirst
{
    int Value = 0;

    if (q->iCount == 0)
    {
        printf("Queue is empty\n");
        return -1;
    }
    else if (q->iCount == 1)
    {
        Value = q->First->data;
        free(q->First);
        q->First = NULL;
    }
    else
    {
        Value = q->First->data;
        PNODE temp = q->First;
        q->First = q->First->next;
        free(temp);
    }

    q->iCount--;

    return Value;
}

void Display(QUEUE *q)
{
    if (q->First == NULL)
    {
        printf("Nothing to display as Queue is empty\n");
        return;
    }

    printf("Elements of Queue are : \n");

    PNODE temp = q->First;
    while (temp != NULL)
    {
        printf("%d\n", temp->data);
        temp = temp->next;
    }
}

int Count(QUEUE *q)
{
    return q->iCount;
}

// Main function
int main()
{
    int iChoice = 1;
    int iValue = 0;
    int iRet = 0;

    QUEUE obj;
    Initialize(&obj);

    printf("-----------------------------------------------------\n");
    printf("Dynamic Implementation of Queue in C\n");
    printf("-----------------------------------------------------\n");

    while (iChoice != 0)
    {
        printf("-----------------------------------------------------\n");
        printf("1 : Insert element into Queue\n");
        printf("2 : Remove element from the Queue\n");
        printf("3 : Display elements from Queue\n");
        printf("4 : Count number of elements from Queue\n");
        printf("0 : Terminate the application\n");
        printf("-----------------------------------------------------\n");

        printf("Please enter the option : \n");
        scanf("%d", &iChoice);

        switch (iChoice)
        {
        case 1:
            printf("Enter the element that you want to insert\n");
            scanf("%d", &iValue);
            Enqueue(&obj, iValue);
            break;

        case 2:
            iRet = Dequeue(&obj);
            if (iRet != -1)
            {
                printf("Removed element from Queue is : %d\n", iRet);
            }
            break;

        case 3:
            Display(&obj);
            break;

        case 4:
            iRet = Count(&obj);
            printf("Number of elements are : %d\n", iRet);
            break;

        case 0:
            printf("Thank you for using the application\n");
            break;

        default:
            printf("Please enter valid option\n");
            break;
        } // End of switch
    }     // End of while

    return 0;
}

File: Queue.java
import java.util.Scanner;

// Node class
class Node {
    int data;
    Node next;

    Node(int value) {
        data = value;
        next = null;
    }
}

// Queue class
class Queue {
    private Node first;
    private int count;

    public Queue() {
        first = null;
        count = 0;
    }

    // InsertLast
    public void enqueue(int no) {
        Node newn = new Node(no);

        if (first == null) {
            first = newn;
        } else {
            Node temp = first;
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newn;
        }
        count++;
    }

    // DeleteFirst
    public int dequeue() {
        int value = 0;

        if (count == 0) {
            System.out.println("Queue is empty");
            return -1;
        } else if (count == 1) {
            value = first.data;
            first = null;
        } else {
            value = first.data;
            Node temp = first;
            first = first.next;
            temp = null;  // GC will clean up
        }

        count--;
        return value;
    }

    public void display() {
        if (first == null) {
            System.out.println("Nothing to display as Queue is empty");
            return;
        }

        System.out.println("Elements of Queue are : ");
        Node temp = first;
        while (temp != null) {
            System.out.println(temp.data);
            temp = temp.next;
        }
    }

    public int count() {
        return count;
    }
}

// Main class
public class QueueDemo {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int iChoice = 1;
        int iValue = 0;
        int iRet = 0;

        Queue obj = new Queue();

        System.out.println("-----------------------------------------------------");
        System.out.println("Dynamic Implementation of Queue in Java");
        System.out.println("-----------------------------------------------------");

        while (iChoice != 0) {
            System.out.println("-----------------------------------------------------");
            System.out.println("1 : Insert element into Queue");
            System.out.println("2 : Remove element from the Queue");
            System.out.println("3 : Display elements from Queue");
            System.out.println("4 : Count number of elements from Queue");
            System.out.println("0 : Terminate the application");
            System.out.println("-----------------------------------------------------");

            System.out.print("Please enter the option : ");
            iChoice = sc.nextInt();

            switch (iChoice) {
                case 1:
                    System.out.print("Enter the element that you want to insert: ");
                    iValue = sc.nextInt();
                    obj.enqueue(iValue);
                    break;

                case 2:
                    iRet = obj.dequeue();
                    if (iRet != -1) {
                        System.out.println("Removed element from Queue is : " + iRet);
                    }
                    break;

                case 3:
                    obj.display();
                    break;

                case 4:
                    iRet = obj.count();
                    System.out.println("Number of elements are : " + iRet);
                    break;

                case 0:
                    System.out.println("Thank you for using the application");
                    break;

                default:
                    System.out.println("Please enter valid option");
                    break;
            }
        }

        sc.close();
    }
}
